README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# giscoR <a href='https://ropengov.github.io/giscoR/'><img src="man/figures/logo.png" align="right" height="139"/></a>

<!-- badges: start -->

[![rOG-badge](https://ropengov.github.io/rogtemplate/reference/figures/ropengov-badge.svg)](https://ropengov.org/)
[![CRAN
status](https://www.r-pkg.org/badges/version/giscoR)](https://CRAN.R-project.org/package=giscoR)
[![CRAN
results](https://badges.cranchecks.info/worst/giscoR.svg)](https://cran.r-project.org/web/checks/check_results_giscoR.html)
[![Downloads](https://cranlogs.r-pkg.org/badges/giscoR)](https://CRAN.R-project.org/package=giscoR)
[![r-universe](https://ropengov.r-universe.dev/badges/giscoR)](https://ropengov.r-universe.dev/giscoR)
[![R build
status](https://github.com/rOpenGov/giscoR/workflows/R-CMD-check/badge.svg)](https://github.com/rOpenGov/giscoR/actions)
[![codecov](https://codecov.io/gh/ropengov/giscoR/branch/main/graph/badge.svg)](https://app.codecov.io/gh/ropengov/giscoR)
[![DOI](https://img.shields.io/badge/DOI-10.5281/zenodo.4317946-blue)](https://doi.org/10.5281/zenodo.4317946)
[![Project Status:
Active](https://www.repostatus.org/badges/latest/active.svg)](https://www.repostatus.org/#active)
[![status](https://tinyverse.netlify.com/badge/giscoR)](https://CRAN.R-project.org/package=giscoR)

<!-- badges: end -->

[giscoR](https://ropengov.github.io/giscoR//) is an API package that
helps to retrieve data from [Eurostat - GISCO (the Geographic
Information System of the
COmmission)](https://ec.europa.eu/eurostat/web/gisco). It also provides
some lightweight data sets ready to use without downloading.

GISCO [(FAQ)](https://ec.europa.eu/eurostat/web/gisco/faq) is a
geospatial open data repository including several data sets as
countries, coastal lines, labels or [NUTS
levels](https://ec.europa.eu/eurostat/web/regions-and-cities/overview).
The data sets are usually provided at several resolution levels
(60M/20M/10M/03M/01M) and in 3 different projections (4326/3035/3857).

Note that the package does not provide metadata on the downloaded files,
the information is available on the [API
webpage](https://gisco-services.ec.europa.eu/distribution/v2/).

Full site with examples and vignettes on
<https://ropengov.github.io/giscoR/>

## Installation

Install `giscoR` from
[**CRAN**](https://CRAN.R-project.org/package=giscoR):

``` r
install.packages("giscoR")
```

You can install the developing version of `giscoR` with:

``` r
library(remotes)
install_github("rOpenGov/giscoR")
```

Alternatively, you can install `giscoR` using the
[r-universe](https://ropengov.r-universe.dev/giscoR):

``` r
install.packages("giscoR",
  repos = c("https://ropengov.r-universe.dev", "https://cloud.r-project.org")
)
```

## Usage

This script highlights some features of `giscoR`:

``` r
library(giscoR)
library(sf)
library(dplyr)

# Different resolutions
DNK_res60 <- gisco_get_countries(resolution = "60", country = "DNK") %>%
  mutate(res = "60M")
DNK_res20 <-
  gisco_get_countries(resolution = "20", country = "DNK") %>%
  mutate(res = "20M")
DNK_res10 <-
  gisco_get_countries(resolution = "10", country = "DNK") %>%
  mutate(res = "10M")
DNK_res03 <-
  gisco_get_countries(resolution = "03", country = "DNK") %>%
  mutate(res = "03M")


DNK_all <- bind_rows(DNK_res60, DNK_res20, DNK_res10, DNK_res03)

# Plot ggplot2

library(ggplot2)

ggplot(DNK_all) +
  geom_sf(fill = "tomato") +
  facet_wrap(vars(res)) +
  theme_minimal()
```

<img src="https://raw.githubusercontent.com/ropengov/giscoR/main/img/README-example-1.png" width="100%" />

``` r


# Labels and Lines available

labs <- gisco_get_countries(
  spatialtype = "LB",
  region = "Africa",
  epsg = "3857"
)

coast <- gisco_get_countries(
  spatialtype = "COASTL",
  epsg = "3857"
)

# For zooming
afr_bbox <- st_bbox(labs)

ggplot(coast) +
  geom_sf(col = "deepskyblue4", linewidth = 3) +
  geom_sf(data = labs, fill = "springgreen4", col = "darkgoldenrod1", size = 5, shape = 21) +
  coord_sf(
    xlim = afr_bbox[c("xmin", "xmax")],
    ylim = afr_bbox[c("ymin", "ymax")]
  )
```

<img src="https://raw.githubusercontent.com/ropengov/giscoR/main/img/README-example-2.png" width="100%" />

### Labels

An example of a labeled map using `ggplot2`:

``` r
ITA <- gisco_get_nuts(country = "Italy", nuts_level = 1)

ggplot(ITA) +
  geom_sf() +
  geom_sf_text(aes(label = NAME_LATN)) +
  theme(axis.title = element_blank())
```

<img src="https://raw.githubusercontent.com/ropengov/giscoR/main/img/README-labels-1.png" width="100%" />

### Thematic maps

An example of a thematic map plotted with the `ggplot2` package. The
information is extracted via the `eurostat` package. We would follow the
fantastic approach presented by [Milos
Popovic](https://milospopovic.net/) on [this
post](https://milospopovic.net/how-to-make-choropleth-map-in-r/):

We start by extracting the corresponding geographic data:

``` r
# Get shapes
nuts3 <- gisco_get_nuts(
  year = "2016",
  epsg = "3035",
  resolution = "3",
  nuts_level = "3"
)

# Group by NUTS by country and convert to lines
country_lines <- nuts3 %>%
  group_by(
    CNTR_CODE
  ) %>%
  summarise(n = n()) %>%
  st_cast("MULTILINESTRING")
```

We now download the data from Eurostat:

``` r
# Use eurostat
library(eurostat)

popdens <- get_eurostat("demo_r_d3dens") %>%  filter(time == "2018-01-01")
  
```

By last, we merge and manipulate the data for creating the final plot:

``` r
# Merge data
nuts3.sf <- nuts3 %>%
  left_join(popdens, by = c("NUTS_ID" = "geo"))


# Breaks and labels

br <- c(0, 25, 50, 100, 200, 500, 1000, 2500, 5000, 10000, 30000)

nuts3.sf <- nuts3.sf %>%
  mutate(values_cut = cut(values, br, dig.lab = 5))

labs_plot <- prettyNum(br[-1], big.mark = ",")

# Palette
pal <- hcl.colors(length(br) - 1, "Lajolla")


# Plot

ggplot(nuts3.sf) +
  geom_sf(aes(fill = values_cut), linewidth = 0, color = NA, alpha = 0.9) +
  geom_sf(data = country_lines, col = "black", linewidth = 0.1) +
  # Center in Europe: EPSG 3035
  coord_sf(
    xlim = c(2377294, 7453440),
    ylim = c(1313597, 5628510)
  ) +
  labs(
    title = "Population density in 2018",
    subtitle = "NUTS-3 level",
    caption = paste0(
      "Source: Eurostat, ", gisco_attributions(),
      "\nBased on Milos Popovic: https://milospopovic.net/how-to-make-choropleth-map-in-r/"
    )
  ) +
  scale_fill_manual(
    name = "people per sq. kilometer",
    values = pal,
    labels = labs_plot,
    drop = FALSE,
    guide = guide_legend(
      direction = "horizontal",
      keyheight = 0.5,
      keywidth = 2.5,
      title.position = "top",
      title.hjust = 0.5,
      label.hjust = .5,
      nrow = 1,
      byrow = TRUE,
      reverse = FALSE,
      label.position = "bottom"
    )
  ) +
  theme_void() +
  # Theme
  theme(
    plot.title = element_text(
      size = 20, color = pal[length(pal) - 1],
      hjust = 0.5, vjust = -6
    ),
    plot.subtitle = element_text(
      size = 14,
      color = pal[length(pal) - 1],
      hjust = 0.5, vjust = -10, face = "bold"
    ),
    plot.caption = element_text(
      size = 9, color = "grey60",
      hjust = 0.5, vjust = 0,
      margin = margin(t = 5, b = 10)
    ),
    legend.text = element_text(
      size = 10,
      color = "grey20"
    ),
    legend.title = element_text(
      size = 11,
      color = "grey20"
    ),
    legend.position = "bottom"
  )
```

<img src="https://raw.githubusercontent.com/ropengov/giscoR/main/img/README-thematic-1.png" width="100%" />

## A note on caching

Some data sets (as Local Administrative Units - LAU, or high-resolution
files) may have a size larger than 50MB. You can use `giscoR` to create
your own local repository at a given local directory passing the
following function:

``` r
gisco_set_cache_dir("./path/to/location")
```

You can also download manually the files (`.geojson` format) and store
them on your local directory.

## Recommended packages

### API data packages

- `eurostat` package (<https://ropengov.github.io/eurostat/>). This is
  an API package that provides access to open data from Eurostat.

### Plotting `sf` objects

Some packages recommended for visualization are:

- [`tmap`](https://r-tmap.github.io/tmap/)
- [`ggplot2`](https://github.com/tidyverse/ggplot2) +
  [`ggspatial`](https://github.com/paleolimbot/ggspatial)
- [`mapsf`](https://riatelab.github.io/mapsf/)
- [`leaflet`](https://rstudio.github.io/leaflet/)

## Contribute

Check the GitHub page for [source
code](https://github.com/rOpenGov/giscoR/).

Contributions are very welcome:

- [Use issue tracker](https://github.com/rOpenGov/giscoR/issues) for
  feedback and bug reports.
- [Send pull requests](https://github.com/rOpenGov/giscoR/)
- [Star us on the GitHub page](https://github.com/rOpenGov/giscoR)

## Citation

To cite ‘giscoR’ in publications use:

Hernangomez D (2023). giscoR: Download Map Data from GISCO API -
Eurostat. <https://doi.org/10.5281/zenodo.4317946>,
<https://ropengov.github.io/giscoR/>

A BibTeX entry for LaTeX users is

    @Manual{R-giscoR,
      title = {{giscoR}: Download Map Data from GISCO API - Eurostat},
      doi = {10.5281/zenodo.4317946},
      author = {Diego Hernangómez},
      year = {2023},
      version = {0.3.5},
      url = {https://ropengov.github.io/giscoR/},
      abstract = {Tools to download data from the GISCO (Geographic Information System of the Commission) Eurostat database <https://ec.europa.eu/eurostat/web/gisco>. Global and European map data available. This package is in no way officially related to or endorsed by Eurostat.},
    }

## Copyright notice

*From GISCO \> Geodata \> Reference data \> Administrative Units /
Statistical Units*

> When data downloaded from this page is used in any printed or
> electronic publication, in addition to any other provisions applicable
> to the whole Eurostat website, data source will have to be
> acknowledged in the legend of the map and in the introductory page of
> the publication with the following copyright notice:
>
> EN: © EuroGeographics for the administrative boundaries
>
> FR: © EuroGeographics pour les limites administratives
>
> DE: © EuroGeographics bezüglich der Verwaltungsgrenzen
>
> For publications in languages other than English, French or German,
> the translation of the copyright notice in the language of the
> publication shall be used.

If you intend to use the data commercially, please contact
EuroGeographics for information regarding their license agreements.

## Disclaimer

This package is in no way officially related to or endorsed by Eurostat.


NEWS.md
# giscoR 0.3.5

-   Review examples to avoid CRAN errors and notes.

-   New helper function: `gisco_detect_cache_dir()`.

-   Now the functions fail gracefully with an informative message, instead of an
    error, and return `NULL`.

# giscoR 0.3.4

-   Update tests and documentation.

# giscoR 0.3.3

-   Fix broken urls on `gisco_get_healthcare()` (#51).

# giscoR 0.3.2

-   Fix HTML5 issue as requested by CRAN.

# giscoR 0.3.1

-   Add `Copyright` on DESCRIPTION.
-   Add `lwgeom` on `Suggests`.
-   `gisco_get_airports()` and `gisco_get_ports()`:
    -   Only year available is 2013.
    -   Now information is downloaded instead of using internal data.
-   New function: `gisco_get_postalcodes()`.
-   Update `gisco_db`.

# giscoR 0.3.0

-   Now **giscoR** is part of [rOpenGov](https://ropengov.org/). Repo has been
    transferred.
-   Caching improvements: new function `gisco_set_cache_dir()` based on
    `rappdirs::user_cache_dir()`. Now the `cache_dir` path is stored and it is
    not necessary to set it up again on a new session. Also added
    `gisco_clear_cache()`.
-   Fix an error when `cache = FALSE`. Now files are loaded instead throwing an
    error.
-   New tests with `testthat`.
-   Update on docs. New examples
-   Refactor documents and codes for the previous `gisco_get` doc.
-   Add `eurostat` package to `Suggests`.
-   `lwgeom` dependency removed.
-   Update internal grid object.
-   `tmap` package replaced by `ggplot2` on vignettes and examples.

# giscoR 0.2.4

-   New `eu` field on `giscoR::gisco_countrycode`.
-   Fix typos on documentation
-   Include vignette on the package
-   Move docs to markdown/roxygen
-   `lwgeom` moved to Import field.
-   `cartography` package replaced by `tmap` on vignettes.

# giscoR 0.2.3

-   Update on docs
-   Release for DOI

# giscoR 0.2.2

-   Remove vignette

# giscoR 0.2.1

-   Remove CRAN notes.
-   Improve docs.
-   Fix CRAN checks.

# giscoR 0.2.0

-   Remove `colorspace` as dependency.
-   Bump R minimal version to 3.6.0.
-   Change order on parameters for `gisco_get()` functions.
-   Rewriting of internal functions and utils.
-   Add `verbose` parameter to functions.
-   Rewriting of `giscoR::gisco_db`.
-   Functions added:
    -   `gisco_bulk_download()`
    -   `gisco_check_access()`
    -   `gisco_get_airports()`
    -   `gisco_get_grid()`
    -   `gisco_get_ports()`
    -   `gisco_get_units()`
-   Now `gisco_get_countries()` and `gisco_get_nuts()` uses `gisco_get_units()`
    for individual files, making the call much faster.

# giscoR 0.1.1

-   Added `giscoR::tgs00026` dataset.
-   Remove `eurostat` dependency.

# giscoR 0.1.0

-   First stable release.


R/data.R
#' GISCO database
#'
#' Database with the list of files that the package can load.
#'
#' @concept dataset
#'
#' @name gisco_db
#'
#' @docType data
#'
#' @format A data frame
#'
#' @details This dataframe is used to check the validity of the API calls.
#'
#' @source GISCO API `datasets.json`.
#' @keywords internal
#' @examples
#'
#' data(gisco_db)
NULL

#' World countries `POLYGON` object
#'
#' @family dataset
#'
#' @name gisco_countries
#'
#' @description A `sf` object including all
#' countries as provided by GISCO (2016 version).
#'
#' @format
#' A `MULTIPOLYGON` data frame (resolution: 1:20million, EPSG:4326) object
#' with `r nrow(giscoR::gisco_countries)` rows and 7 variables:
#'   * **id**: row ID
#'   * **CNTR_NAME**: Official country name on local language
#'   * **ISO3_CODE**: ISO 3166-1 alpha-3 code of each country, as provided by
#'   GISCO
#'   * **CNTR_ID**: Country ID
#'   * **NAME_ENGL**: Country name in English
#'   * **FID**: FID
#'   * **geometry**: geometry field
#'
#' @examples
#'
#' cntry <- gisco_countries
#' GBR <- subset(cntry, ISO3_CODE == "GBR")
#'
#' library(ggplot2)
#'
#' ggplot(GBR) +
#'   geom_sf(color = "red3", fill = "blue4") +
#'   theme_void()
#' @source
#' [CNTR_RG_20M_2016_4326.geojson](https://gisco-services.ec.europa.eu/distribution/v2/countries/geojson/) file.
#'
#' @docType data
#'
#' @seealso [gisco_get_countries()]
#'
#' @encoding UTF-8
NULL

#' World coastal lines `POLYGON` object
#'
#' A `sf` object as provided by GISCO (2016 version).
#'
#' @family dataset
#'
#' @name gisco_coastallines
#'
#' @format
#' A `POLYGON` data frame (resolution: 1:20million, EPSG:4326) object with
#' 3 variables:
#'   * **FID**
#'   * **COAS_ID**
#'   * **geometry**: geometry field
#'
#' @source
#' [COAS_RG_20M_2016_4326.geojson](https://gisco-services.ec.europa.eu/distribution/v2/coas/geojson/) file.
#'
#' @docType data
#'
#' @seealso [gisco_get_coastallines()]
#'
#' @examples
#'
#' coasts <- gisco_coastallines
#'
#' library(ggplot2)
#'
#' ggplot(coasts) +
#'   geom_sf(color = "blue", fill = "blue", alpha = 0.2) +
#'   # Zoom on Oceania
#'   coord_sf(
#'     xlim = c(96, 179),
#'     ylim = c(-51, 11)
#'   ) +
#'   theme_minimal() +
#'   theme(
#'     plot.background = element_rect(
#'       fill = "black",
#'       color = "black"
#'     ),
#'     panel.grid = element_blank(),
#'     axis.text = element_text(colour = "grey90")
#'   )
NULL

#' All NUTS `POLYGON` object
#'
#' A `sf` object including all NUTS levels as provided by GISCO (2016 version).
#'
#' @family dataset
#'
#' @name gisco_nuts
#'
#' @format
#' A `POLYGON` data frame (resolution: 1:20million, EPSG:4326) object with
#' `r prettyNum(nrow(giscoR::gisco_nuts), big.mark = ",")` rows and
#' 11 variables:
#'   * **id**: row ID
#'   * **COAST_TYPE**: COAST_TYPE
#'   * **MOUNT_TYPE**: MOUNT_TYPE
#'   * **NAME_LATN**: Name on Latin characters
#'   * **CNTR_CODE**: Eurostat Country code
#'   * **FID**: FID
#'   * **NUTS_ID**: NUTS identifier
#'   * **NUTS_NAME**: NUTS name on local alphabet
#'   * **LEVL_CODE**: NUTS level code (0,1,2,3)
#'   * **URBN_TYPE**: URBN_TYPE
#'   * **geometry**: geometry field
#'
#' @source
#' [NUTS_RG_20M_2016_4326.geojson](https://gisco-services.ec.europa.eu/distribution/v2/nuts/geojson/) file.
#'
#' @docType data
#'
#' @seealso [gisco_get_nuts()]
#'
#' @examples
#'
#' nuts <- gisco_nuts
#'
#' italy <- subset(nuts, CNTR_CODE == "IT" & LEVL_CODE == 3)
#'
#' library(ggplot2)
#'
#' ggplot(italy) +
#'   geom_sf()
#' @encoding UTF-8
NULL


#' Dataframe with different country code schemes and world regions
#'
#' @name gisco_countrycode
#'
#' @family dataset
#'
#' @description
#' A dataframe containing conversions between different country
#' code schemes (Eurostat/ISO2 and 3) as well as geographic regions as
#' provided by the World Bank and the UN (M49). This dataset
#' is extracted from **countrycode** package.
#'
#' @format
#' A data frame object with 249 rows and 12 variables:
#'   * **CNTR_CODE**: Eurostat code of each country
#'   * **iso2c**: ISO 3166-1 alpha-2 code of each country
#'   * **ISO3_CODE**: ISO 3166-1 alpha-3 code of each country
#'   * **iso.name.en**: ISO English short name
#'   * **cldr.short.en**: English short name as provided by the Unicode Common
#'   Locale Data Repository
#'   <https://cldr.unicode.org/translation/displaynames/countryregion-territory-names>
#'   * **continent**: As provided by the World Bank
#'   * **un.region.code**: Numeric region code UN (M49)
#'   * **un.region.name**: Region name UN (M49)
#'   * **un.regionintermediate.code**: Numeric intermediate Region
#'    code UN (M49)
#'   * **un.regionintermediate.name**: Intermediate Region name UN (M49)
#'   * **un.regionsub.code**: Numeric sub-region code UN (M49)
#'   * **un.regionsub.name**: Sub-Region name UN (M49)
#'   * **eu**: Logical indicating if the country belongs to the European
#'   Union as per February 2021.
#'
#' @examples
#'
#' data(gisco_countrycode)
#' @source [countrycode::codelist] **v1.2.0**.
#'
#' @seealso [gisco_get_countries()],
#'  [countrycode::codelist], [countrycode::countrycode-package]
#'
#' @docType data
NULL

#' Disposable income of private households by NUTS 2 regions
#'
#' @name tgs00026
#'
#' @family dataset
#'
#' @source <https://ec.europa.eu/eurostat>, extracted on 2020-10-27
#'
#' @description
#' The disposable income of private households is the balance
#' of primary income (operating surplus/mixed income plus compensation of
#' employees plus property income received minus property income paid) and
#' the redistribution of income in cash. These transactions comprise social
#' contributions paid, social benefits in cash received, current taxes on
#' income and wealth paid, as well as other current transfers. Disposable
#' income does not include social transfers in kind coming from public
#' administrations or non-profit institutions serving households.
#' @format
#' data_frame:
#'   * **geo**: NUTS2 identifier
#'   * **time**: reference year (2007 to 2018)
#'   * **values**: value in euros
#'
#' @examples
#'
#' data(tgs00026)
#' @docType data
NULL
